packages/playwright-test/src/matchers/expect.ts

```diff
@@ -106,28 +107,57 @@ export const printReceivedStringContainExpectedResult = (
 
 // #endregion
 
-type ExpectMessageOrOptions = undefined | string | { message?: string, timeout?: number, intervals?: number[] };
+type ExpectMessage = string | { message?: string };
 
-function createExpect(actual: unknown, messageOrOptions: ExpectMessageOrOptions, isSoft: boolean, isPoll: boolean, generator?: Generator): any {
-  return new Proxy(expectLibrary(actual), new ExpectMetaInfoProxyHandler(messageOrOptions, isSoft, isPoll, generator));
+function createMatchers(actual: unknown, info: ExpectMetaInfo): any {
+  return new Proxy(expectLibrary(actual), new ExpectMetaInfoProxyHandler(info));
 }
 
-export const expect: Expect = new Proxy(expectLibrary, {
-  apply: function(target: any, thisArg: any, argumentsList: [actual: unknown, messageOrOptions: ExpectMessageOrOptions]) {
-    const [actual, messageOrOptions] = argumentsList;
-    return createExpect(actual, messageOrOptions, false /* isSoft */, false /* isPoll */);
-  }
-});
+function createExpect(info: ExpectMetaInfo) {
+  const expect: Expect = new Proxy(expectLibrary, {
+    apply: function(target: any, thisArg: any, argumentsList: [unknown, ExpectMessage?]) {
+      const [actual, messageOrOptions] = argumentsList;
+      const message = isString(messageOrOptions) ? messageOrOptions : messageOrOptions?.message || info.message;
+      const newInfo = { ...info, message };
+      if (newInfo.isPoll)
+        newInfo.generator = actual as any;
+      return createMatchers(actual, newInfo);
+    }
+  });
 
-expect.soft = (actual: unknown, messageOrOptions: ExpectMessageOrOptions) => {
-  return createExpect(actual, messageOrOptions, true /* isSoft */, false /* isPoll */);
-};
+  expect.soft = (actual: unknown, messageOrOptions?: ExpectMessage) => {
+    return expect.configure({ soft: true })(actual, messageOrOptions) as any;
+  };
 
-expect.poll = (actual: unknown, messageOrOptions: ExpectMessageOrOptions) => {
-  if (typeof actual !== 'function')
-    throw new Error('`expect.poll()` accepts only function as a first argument');
-  return createExpect(actual, messageOrOptions, false /* isSoft */, true /* isPoll */, actual as any);
-};
+  expect.poll = (actual: unknown, messageOrOptions?: ExpectMessage & { timeout?: number, intervals?: number[] }) => {
+    if (typeof actual !== 'function')
```

Comment:
I still think we should move this somewhere deeper, because `configure({ poll })` now misses this logic.